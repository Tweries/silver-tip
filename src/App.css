.App {
  background-color: #f5f5f5;
  border: 1px solid gray;
  margin: 32px auto;
  padding: 16px;
  text-align: center;
  width: 400px;
}

.App__button,
.App__h1,
.App__input,
.App__textarea {
  background-color: transparent;
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.App__input,
.App__textarea {
  border: 0 solid gray;
  border-bottom-width: 1px;
  font-size: 16px;
  margin-bottom: 8px;
  padding: 8px;
}

.App__button {
  background-color: blue;
  border-width: 0;
  color: white;
  font-size: 16px;
  font-weight: bold;
  margin: 8px auto 0 auto;
  padding: 8px 16px;
}

.App__form {
  display: flex;
  flex-direction: column;
}
