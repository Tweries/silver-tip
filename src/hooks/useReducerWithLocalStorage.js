import { useReducer } from 'react';
import useLocalStorage from './useLocalStorage';

function useReducerWithLocalStorage({ initializerArg, key, reducer }) {
  const [localStorageState, setLocalStorageState] = useLocalStorage(
    key,
    JSON.stringify(initializerArg)
  );

  return useReducer(
    (state, action) => {
      const newState = reducer(state, action);
      setLocalStorageState(JSON.stringify(newState));
      return newState;
    },
    {
      ...JSON.parse(localStorageState)
    }
  );
}

export default useReducerWithLocalStorage;
