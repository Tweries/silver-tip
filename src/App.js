import React, { useReducer } from 'react';
import './App.css';

const CHANGE_BIO = 'CHANGE_BIO';
const CHANGE_NAME = 'CHANGE_NAME';
const CHANGE_URL = 'CHANGE_URL';

function App() {
  function reducer(state, action) {
    console.log(action);
    switch (action.type) {
      case CHANGE_BIO:
        return { ...state, bio: action.value };
      case CHANGE_NAME:
        return { ...state, name: action.value };
      case CHANGE_URL:
        return { ...state, url: action.value };
      default:
        return state;
    }
  }

  const initializerArgs = { bio: '', name: '', url: '' };

  const [{ bio }, dispatch] = useReducer(reducer, initializerArgs);

  return (
    <div className="App">
      <h1 className="App__h1">Public profile</h1>
      <form className="App__form">
        <input
          className="App__input"
          name="name"
          onChange={(e) =>
            dispatch({ type: CHANGE_NAME, value: e.target.value })
          }
          placeholder="Name"
          type="text"
        />
        <textarea
          className="App__textarea"
          cols="40"
          name="bio"
          onChange={(e) =>
            dispatch({ type: CHANGE_BIO, value: e.target.value })
          }
          placeholder="Bio"
          rows="10"
          value={bio}
        />
        <input
          className="App__input"
          name="url"
          onChange={(e) =>
            dispatch({ type: CHANGE_URL, value: e.target.value })
          }
          placeholder="URL"
          type="text"
        />
      </form>
    </div>
  );
}

export default App;
