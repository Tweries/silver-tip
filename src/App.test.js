import { render } from '@testing-library/react';
import React from 'react';
import useReducerWithLocalStorage from './hooks/useReducerWithLocalStorage';
import App from './App';

jest.mock('./hooks/useReducerWithLocalStorage');

describe('<App />', () => {
  useReducerWithLocalStorage.mockImplementation(() => [
    {
      bio: 'CEO of Microsoft',
      name: 'Satya Nadella',
      url: 'https://www.microsoft.com/en-us/'
    },
    jest.fn()
  ]);

  it('to match snapshot', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
