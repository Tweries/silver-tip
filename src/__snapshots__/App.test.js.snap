// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> to match snapshot 1`] = `
<div>
  <div
    class="App"
  >
    <h1
      class="App__h1"
    >
      Public profile
    </h1>
    <form
      class="App__form"
    >
      <input
        class="App__input"
        name="name"
        placeholder="Name"
        type="text"
        value="Satya Nadella"
      />
      <textarea
        class="App__textarea"
        cols="40"
        name="bio"
        placeholder="Bio"
        rows="10"
      >
        CEO of Microsoft
      </textarea>
      <input
        class="App__input"
        name="url"
        placeholder="URL"
        type="text"
        value="https://www.microsoft.com/en-us/"
      />
      <button
        class="App__button"
        type="button"
      >
        Clear
      </button>
    </form>
  </div>
</div>
`;
